### PR TITLE
feat: upgrade DataFusion, Arrow, PyO3, ObjectStore

### PR DESCRIPTION
**Pull request by wjones127**
BREAKING CHANGE:

* Removed access to `MultipartId`, `cleanup_partial_writes`. Uploads are now automatically cleaned up.
  * `FragmentWriteProgress` no longer has any `multipart_id` argument.
* `CommitHandler` and `ManifestWriter` APIs now take a Lance `ObjectStore` rather than `object_store::ObjectStore`.

New features:

* For S3 and Azure (these features were already implemented for GCS):
  * Max object size is now 2.5 TB, up from 50GB.
  * Retries are now performed for `Connection reset` network errors.
  * Reduced number of IOPS for writing small objects (<5MB)

### CODE REVIEW

File: rust/lance-io/src/object_writer.rs
Comment (wjones127) [inline rust/lance-io/src/object_writer.rs:1]: Most of this was ported from the now deleted `gcs_wrapper.rs`

File: rust/lance-io/src/object_writer.rs [hunk 1]
```diff
@@ -58,44 +274,185 @@ impl ObjectWriter {
...
+            // Take ownership of the state.
+            let state = std::mem::replace(&mut self.state, UploadState::Done);
+            if let UploadState::InProgress { mut upload, .. } = state {
+                tokio::task::spawn(async move {
+                    let _ = upload.abort().await;
+                });
+            }
+        }
     }
 }
```
Comment (wjones127) [inline rust/lance-io/src/object_writer.rs:290]: This is what ensures partial uploads are cleaned up now. `MultipartId` is no longer exposed in the `object-store` API.

File: Cargo.toml [hunk 2]
```diff
@@ -43,33 +43,33 @@ categories = [
...
-arrow-array = "51.0"
-arrow-buffer = "51.0"
-arrow-cast = "51.0"
-arrow-data = "51.0"
-arrow-ipc = { version = "51.0", features = ["zstd"] }
-arrow-ord = "51.0"
-arrow-row = "51.0"
-arrow-schema = "51.0"
-arrow-select = "51.0"
+arrow = { version = "52.0", optional = false, features = ["prettyprint"] }
```
Comment (BubbleCal) [inline Cargo.toml:62]: that `FixedSizeListArray` offset overflow fix is in `52.1`, but now datafusion hasn't make a release matches that arrow version, so I will upgrade them again after datafusion makes a new release

File: python/python/lance/cleanup.py [hunk 3]
```diff
@@ -1,20 +1,3 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright The Lance Authors
 
-from typing import List, Tuple
```
Comment (BubbleCal) [inline python/python/lance/cleanup.py:4]: can we just delete this file?
it's empty now, will we add something in future?

File: rust/lance-datagen/src/generator.rs [hunk 4]
```diff
@@ -1664,8 +1662,11 @@ pub mod array {
                 TimeUnit::Nanosecond => rand::<DurationNanosecondType>(),
             },
             DataType::Interval(unit) => match unit {
-                IntervalUnit::DayTime => rand::<IntervalDayTimeType>(),
-                IntervalUnit::MonthDayNano => rand::<IntervalMonthDayNanoType>(),
+                // TODO: fix these. In Arrow they changed to have specialized
+                // Native types, which don't support Distribution.
+                // IntervalUnit::DayTime => rand::<IntervalDayTimeType>(),
+                // IntervalUnit::MonthDayNano => rand::<IntervalMonthDayNanoType>(),
+                IntervalUnit::DayTime | IntervalUnit::MonthDayNano => todo!(),
```
Comment (BubbleCal) [inline rust/lance-datagen/src/generator.rs:1669]: I added some utils to do this,
post it here if you would like to use it
https://github.com/lancedb/lance/pull/2589/files#diff-05a6cb2a51514a3883006392e5a3b6990e908be79d42414de787ba5c5b623c87R600

File: rust/lance-io/src/object_store.rs [hunk 5]
```diff
@@ -308,6 +308,11 @@ pub struct ObjectStoreParams {
     pub aws_credentials: Option<AwsCredentialProvider>,
     pub object_store_wrapper: Option<Arc<dyn WrappingObjectStore>>,
     pub storage_options: Option<HashMap<String, String>>,
+    /// Use constant size upload parts for multipart uploads. Only necessary
+    /// for Cloudflare R2, which doesn't support variable size parts. When this
+    /// is false, max upload size is 2.5TB. When this is true, the max size is
```
Comment (BubbleCal) [inline rust/lance-io/src/object_store.rs:313]: Saw in PR body you mentioned it's 2.5GB, guess 2.5TB is the right one?

Comment (wjones127) [reply rust/lance-io/src/object_store.rs:313]: Yeah thanks for the correction.

File: Cargo.toml (hunk 2 above)
Comment (wjones127) [reply Cargo.toml:62]: I think I can just make that upgrade quickly.

File: rust/lance-datagen/src/generator.rs (hunk 4 above)
Comment (wjones127) [reply rust/lance-datagen/src/generator.rs]: Thanks!

File: rust/lance-datagen/src/generator.rs [hunk 6]
```diff
@@ -1663,11 +1717,7 @@ pub mod array {
                 TimeUnit::Microsecond => rand::<DurationMicrosecondType>(),
                 TimeUnit::Nanosecond => rand::<DurationNanosecondType>(),
             },
-            DataType::Interval(unit) => match unit {
-                IntervalUnit::DayTime => rand::<IntervalDayTimeType>(),
-                IntervalUnit::MonthDayNano => rand::<IntervalMonthDayNanoType>(),
-                IntervalUnit::YearMonth => rand::<IntervalYearMonthType>(),
-            },
+            DataType::Interval(unit) => rand_interval(*unit),
```
Comment (westonpace) [inline rust/lance-datagen/src/generator.rs:1720]: Why the change here?

File: rust/lance/src/dataset/cleanup.rs [hunk 7]
```diff
@@ -394,61 +393,20 @@ pub async fn cleanup_old_versions(
...
-/// function to clean them up immediately.
-///
-/// To find partial writes, you can use the
-/// [crate::dataset::progress::WriteFragmentProgress] trait to track which files
-/// have been started but never finished.
-pub async fn cleanup_partial_writes(
-    store: &ObjectStore,
-    base_path: &Path,
-    objects: impl IntoIterator<Item = (&Path, &String)>,
-) -> Result<()> {
```
Comment (westonpace) [inline rust/lance/src/dataset/cleanup.rs:410]: Why get rid of this?  Was it no longer working?  Or no longer needed?

**Review by westonpace — APPROVED** [review]
This looks quite exhausting.  Thanks for the cleanup

File: rust/lance-datagen/src/generator.rs (hunk 6 above)
Comment (wjones127) [reply rust/lance-datagen/src/generator.rs:1720]: I think they changed the underlying native type to a struct rather than i128/i64. Thus `Distribution` was no longer implemented.

https://github.com/apache/arrow-rs/pull/5769

File: rust/lance/src/dataset/cleanup.rs (hunk 7 above)
Comment (wjones127) [reply rust/lance/src/dataset/cleanup.rs:410]: `object_store` no longer exposes the multipart ids. So now users can't collect them and pass them into this function.

Instead, we automatically cleanup on `Drop` for `ObjectWriter`. In case of crashes, users should rely on setting a lifecycle rule to delete old incomplete writes. This is already recommended in our docs.
